src/main/scala/goatrodeo/util/ArtifactWrapper.scala

```diff
@@ -37,26 +71,42 @@ case class FileWrapper(f: File, deleteOnFinalize: Boolean)
 
   override def isFile(): Boolean = f.isFile()
 
-  override def listFiles(): Vector[ArtifactWrapper] =
-    f.listFiles().toVector.filter(!_.getName().startsWith(".")).map(FileWrapper(_, false))
+  override def listFiles(): Vector[ArtifactWrapper] = {
+    // the javadocs for `listFiles` says that if it is not a directory, it returns `null`… we need to guard against that
+    Option(f.listFiles()) match {
+      case Some(files) =>
+        files.toVector
+          .filter(!_.getName().startsWith("."))
+          .map(FileWrapper(_, false))
+      case None => Vector.empty
+    }
+  }
 
-  override def getParentDirectory(): File = f.getAbsoluteFile().getParentFile()
+  override def getParentDirectoryPath(): String =
+    f.getAbsoluteFile().getParentFile().getCanonicalPath()
 
   override def getCanonicalPath(): String = f.getCanonicalPath()
 
-  // override def asFile(): (File, Boolean) = (f, false)
-
   override def isDirectory(): Boolean = f.isDirectory()
 
   override def asStream(): InputStream = BufferedInputStream(FileInputStream(f))
 
   override def name(): String = f.getName()
 
   override def size(): Long = f.length()
+
+  override def fileExtension(): Option[String] = {
+    val path = f.getPath()
+    val ext = path.slice(path.lastIndexWhere(_ == '.') + 1, path.length)
```

Comment:
For a file like *foo.tar.gz*, is `gz` generally considered to be the extension, or would it be `tar.gz`?

Could there be a use case where we want a list of extensions, e.g.`List("tar", "gz")`?

Reply:
In the original code, only jar/war/ear (and maybe zip) were special cased in order to use the built-in Java Zip reader. All other readers were handled by "does the reader recognize this file?"

I think the right answer here is to point the ISO code at a file. If the ISO code thinks it's an ISO and can return an iterator, then yay. If it can't move to the next container type opener.

Doing stuff based on file extension should be the absolute last resort.